crypto/hpke/hpke.c

```diff
@@ -0,0 +1,1390 @@
+ * that is sent to the recipient, who calls OSSL_HPKE_open().
+ *
+ * @param suite is the suite to be used
+ * @param enclen points to what will be enc length
+ * @param clearlen is the length of plaintext
+ * @param cipherlen points to what will be ciphertext length (including tag)
+ * @return 1 for success, 0 otherwise
+ */
+static int hpke_expansion(OSSL_HPKE_SUITE suite,
+                          size_t *enclen,
+                          size_t clearlen,
+                          size_t *cipherlen)
+{
+    const OSSL_HPKE_AEAD_INFO *aead_info = NULL;
+    const OSSL_HPKE_KEM_INFO *kem_info = NULL;
+
+    if (cipherlen == NULL || enclen == NULL) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    if (hpke_suite_check(suite) != 1) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    aead_info = ossl_HPKE_AEAD_INFO_find_id(suite.aead_id);
+    if (aead_info == NULL) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    *cipherlen = clearlen + aead_info->taglen;
+    kem_info = ossl_HPKE_KEM_INFO_find_id(suite.kem_id);
+    if (kem_info == NULL) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    *enclen = kem_info->Nenc;
+    return 1;
+}
+
+/*
+ * @brief expand and XOR the 64-bit unsigned seq with (nonce) buffer
+ * @param ctx is the HPKE context
+ * @param buf is the buffer for the XOR'd seq and nonce
+ * @param blen is the size of buf
+ * @return 0 for error, otherwise blen
+ */
+static size_t hpke_seqnonce2buf(OSSL_HPKE_CTX *ctx,
+                                unsigned char *buf, size_t blen)
+{
+    size_t i;
+    uint64_t seq_copy;
+
+    if (ctx == NULL || blen < sizeof(seq_copy) || blen != ctx->noncelen)
+        return 0;
+    seq_copy = ctx->seq;
+    memset(buf, 0, blen);
+    for (i = 0; i < sizeof(seq_copy); i++) {
+        buf[blen - i - 1] = seq_copy & 0xff;
+        seq_copy >>= 8;
+    }
+    for (i = 0; i < blen; i++)
+        buf[i] ^= ctx->nonce[i];
+    return blen;
+}
+
+/*
+ * @brief call the underlying KEM to encap
+ * @param ctx is the OSSL_HPKE_CTX
+ * @param enc is a buffer for the sender's ephemeral public value
+ * @param enclen is the size of enc on input, number of octets used on ouptut
+ * @param pub is the recipient's public value
+ * @param publen is the length of pub
+ * @return 1 for success, 0 for error
+ */
+static int hpke_encap(OSSL_HPKE_CTX *ctx, unsigned char *enc, size_t *enclen,
+                      const unsigned char *pub, size_t publen)
+{
+    int erv = 0;
+    OSSL_PARAM params[3], *p = params;
+    size_t lsslen = 0;
+    EVP_PKEY_CTX *pctx = NULL;
+    EVP_PKEY *pkR = NULL;
+    const OSSL_HPKE_KEM_INFO *kem_info = NULL;
+
+    if (ctx == NULL || enc == NULL || enclen == NULL || *enclen == 0
+        || pub == NULL || publen == 0) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
```

Comment:
error is more close to invalid arguments : `ERR_R_PASSED_NULL_PARAMETER` ??

Reply:
good catch

Reply:
ah - now I check, that's an internal API and those checks are currently duplicate checks (I've left 'em in in case future uses of ``hpke_encap()`` emerge where the inputs weren't checked.)
So, what's right there? Could be a) leave as-is, b) remove the redundant checks or c) change as per the suggestion above. I guess I'd go for (a) but happy to go with (b) or (c) if that's better